File: causeway/sdks/go/raceway/trace_context.go
package raceway

import (
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"net/http"
	"strings"

	"github.com/google/uuid"
)

const (
	traceparentHeader  = "traceparent"
	tracestateHeader   = "tracestate"
	racewayClockHeader = "raceway-clock"

	traceparentVersion = "00"
	traceFlags         = "01"
	clockVersionPrefix = "v1;"
)

type ParsedTraceContext struct {
	TraceID      string
	SpanID       string
	ParentSpanID *string
	TraceState   *string
	ClockVector  []CausalityEntry
	Distributed  bool
}

type PropagationResult struct {
	Headers     map[string]string
	ClockVector []CausalityEntry
	ChildSpanID string
}

type racewayClockPayload struct {
	TraceID      string          `json:"trace_id"`
	SpanID       string          `json:"span_id"`
	ParentSpanID string          `json:"parent_span_id"`
	Service      string          `json:"service"`
	Instance     string          `json:"instance"`
	Clock        [][]interface{} `json:"clock"`
}

func ParseIncomingHeaders(headers http.Header, serviceName, instanceID string) ParsedTraceContext {
	traceID := uuid.New().String()
	var parentSpanID *string
	var traceState *string
	distributed := false

	if raw := headers.Get(traceparentHeader); raw != "" {
		if parsedTrace, ok := parseTraceparent(raw); ok {
			traceID = parsedTrace.traceID
			parentSpanID = parsedTrace.parentSpanID
			distributed = true
		}
	}

	clockVector := []CausalityEntry{}
	if raw := headers.Get(racewayClockHeader); raw != "" {
		if parsedClock, ok := parseRacewayClock(raw); ok {
			if parsedClock.traceID != "" {
				traceID = parsedClock.traceID
			}
			if parsedClock.parentSpanID != nil {
				parentSpanID = parsedClock.parentSpanID
			}
			clockVector = parsedClock.clock
			distributed = true
		}
	}

	if raw := headers.Get(tracestateHeader); raw != "" {
		traceState = &raw
	}

	component := clockComponent(serviceName, instanceID)
	hasComponent := false
	for _, entry := range clockVector {
		if entry.Component() == component {
			hasComponent = true
			break
		}
	}
	if !hasComponent {
		clockVector = append(clockVector, NewCausalityEntry(component, 0))
	}

	return ParsedTraceContext{
		TraceID:      traceID,
		SpanID:       generateSpanID(),
		ParentSpanID: parentSpanID,
		TraceState:   traceState,
		ClockVector:  clockVector,
		Distributed:  distributed,
	}
}

func BuildPropagationHeaders(traceID, currentSpanID string, traceState *string, clockVector []CausalityEntry, serviceName, instanceID string) PropagationResult {
	nextVector := incrementClockVector(clockVector, serviceName, instanceID)
	childSpanID := generateSpanID()

	traceparent := strings.Join([]string{
		traceparentVersion,
		uuidToTraceparent(traceID),
		childSpanID,
		traceFlags,
	}, "-")

	payload := map[string]interface{}{
		"trace_id":       traceID,
		"span_id":        childSpanID,
		"parent_span_id": currentSpanID,
		"service":        serviceName,
		"instance":       instanceID,
		"clock":          encodeClockVector(nextVector),
	}

	payloadJSON, _ := json.Marshal(payload)
	racewayClock := clockVersionPrefix + base64.RawURLEncoding.EncodeToString(payloadJSON)

	headers := map[string]string{
		traceparentHeader:  traceparent,
		racewayClockHeader: racewayClock,
	}
	if traceState != nil {
		headers[tracestateHeader] = *traceState
	}

	return PropagationResult{
		Headers:     headers,
		ClockVector: nextVector,
		ChildSpanID: childSpanID,
	}
}

func incrementClockVector(clockVector []CausalityEntry, serviceName, instanceID string) []CausalityEntry {
	component := clockComponent(serviceName, instanceID)
	next := make([]CausalityEntry, 0, len(clockVector)+1)
	found := false
	for _, entry := range clockVector {
		if entry.Component() == component {
			next = append(next, NewCausalityEntry(entry.Component(), entry.Value()+1))
			found = true
		} else {
			next = append(next, entry)
		}
	}
	if !found {
		next = append(next, NewCausalityEntry(component, 1))
	}
	return next
}

type parsedTraceparent struct {
	traceID      string
	parentSpanID *string
}

func parseTraceparent(value string) (parsedTraceparent, bool) {
	parts := strings.Split(strings.TrimSpace(value), "-")
	if len(parts) != 4 {
		return parsedTraceparent{}, false
	}

	traceIDHex := parts[1]
	spanIDHex := parts[2]
	if len(traceIDHex) != 32 || len(spanIDHex) != 16 {
		return parsedTraceparent{}, false
	}

	if _, err := hex.DecodeString(traceIDHex); err != nil {
		return parsedTraceparent{}, false
	}
	if _, err := hex.DecodeString(spanIDHex); err != nil {
		return parsedTraceparent{}, false
	}

	traceID := traceparentToUUID(traceIDHex)
	parentSpanID := spanIDHex

	return parsedTraceparent{
		traceID:      traceID,
		parentSpanID: &parentSpanID,
	}, true
}

type parsedClock struct {
	traceID      string
	parentSpanID *string
	clock        []CausalityEntry
}

func parseRacewayClock(value string) (parsedClock, bool) {
	if !strings.HasPrefix(value, clockVersionPrefix) {
		return parsedClock{}, false
	}

	encoded := strings.TrimPrefix(value, clockVersionPrefix)
	decoded, err := base64.RawURLEncoding.DecodeString(encoded)
	if err != nil {
		return parsedClock{}, false
	}

	var payload racewayClockPayload
	if err := json.Unmarshal(decoded, &payload); err != nil {
		return parsedClock{}, false
	}

	entries := make([]CausalityEntry, 0, len(payload.Clock))
	for _, item := range payload.Clock {
		if len(item) != 2 {
			continue
		}
		component, ok := item[0].(string)
		if !ok {
			continue
		}
		var valueUint uint64
		switch v := item[1].(type) {
		case float64:
			valueUint = uint64(v)
		case int:
			valueUint = uint64(v)
		case int64:
			valueUint = uint64(v)
		case uint64:
			valueUint = v
		default:
			continue
		}
		entries = append(entries, NewCausalityEntry(component, valueUint))
	}

	var parentSpanID *string
	if payload.SpanID != "" {
		// span_id from sender becomes parent for receiver
		parentSpanID = &payload.SpanID
	}

	return parsedClock{
		traceID:      payload.TraceID,
		parentSpanID: parentSpanID,
		clock:        entries,
	}, true
}

func uuidToTraceparent(value string) string {
	cleaned := strings.ReplaceAll(value, "-", "")
	if len(cleaned) < 32 {
		cleaned = cleaned + strings.Repeat("0", 32-len(cleaned))
	}
	return cleaned[:32]
}

func traceparentToUUID(value string) string {
	return strings.Join([]string{
		value[0:8],
		value[8:12],
		value[12:16],
		value[16:20],
		value[20:32],
	}, "-")
}

func encodeClockVector(clockVector []CausalityEntry) [][]interface{} {
	encoded := make([][]interface{}, 0, len(clockVector))
	for _, entry := range clockVector {
		encoded = append(encoded, []interface{}{entry.Component(), entry.Value()})
	}
	return encoded
}

File: causeway/sdks/python/raceway/trace_context.py
"""Trace context utilities for distributed propagation."""

from __future__ import annotations

import base64
import binascii
import json
import secrets
import uuid
from dataclasses import dataclass
from typing import Mapping, Optional, Tuple, List, Dict

TRACEPARENT_HEADER = "traceparent"
TRACESTATE_HEADER = "tracestate"
RACEWAY_CLOCK_HEADER = "raceway-clock"

TRACEPARENT_VERSION = "00"
TRACE_FLAGS = "01"
CLOCK_VERSION_PREFIX = "v1;"


@dataclass
class ParsedTraceContext:
    trace_id: str
    span_id: str
    parent_span_id: Optional[str]
    tracestate: Optional[str]
    clock_vector: List[Tuple[str, int]]
    distributed: bool


@dataclass
class PropagationHeaders:
    headers: Dict[str, str]
    clock_vector: List[Tuple[str, int]]
    child_span_id: str


def parse_incoming_headers(
    headers: Mapping[str, str],
    *,
    service_name: str,
    instance_id: str,
) -> ParsedTraceContext:
    """Parse inbound HTTP headers into a trace context structure."""
    lower_headers = {k.lower(): v for k, v in headers.items()}

    traceparent_raw = lower_headers.get(TRACEPARENT_HEADER)
    tracestate_raw = lower_headers.get(TRACESTATE_HEADER)
    raceway_clock_raw = lower_headers.get(RACEWAY_CLOCK_HEADER)

    trace_id = str(uuid.uuid4())
    parent_span_id: Optional[str] = None
    distributed = False

    if traceparent_raw:
        parsed = _parse_traceparent(traceparent_raw)
        if parsed:
            trace_id = parsed["trace_id"]
            parent_span_id = parsed["span_id"]
            distributed = True

    clock_vector: List[Tuple[str, int]] = []
    if raceway_clock_raw:
        parsed_clock = _parse_raceway_clock(raceway_clock_raw)
        if parsed_clock:
            clock_vector = parsed_clock["clock"]
            distributed = True
            if parsed_clock["trace_id"]:
                trace_id = parsed_clock["trace_id"]
            if not parent_span_id and parsed_clock["parent_span_id"]:
                parent_span_id = parsed_clock["parent_span_id"]

    component_id = _clock_component(service_name, instance_id)
    if not any(component == component_id for component, _ in clock_vector):
        clock_vector.append((component_id, 0))

    return ParsedTraceContext(
        trace_id=trace_id,
        span_id=_generate_span_id(),
        parent_span_id=parent_span_id,
        tracestate=tracestate_raw,
        clock_vector=clock_vector,
        distributed=distributed,
    )


def build_propagation_headers(
    *,
    trace_id: str,
    current_span_id: str,
    tracestate: Optional[str],
    clock_vector: List[Tuple[str, int]],
    service_name: str,
    instance_id: str,
) -> PropagationHeaders:
    """Build outbound propagation headers and updated clock vector."""
    next_clock_vector = increment_clock_vector(clock_vector, service_name=service_name, instance_id=instance_id)

    child_span_id = _generate_span_id()
    traceparent = "{}-{}-{}-{}".format(
        TRACEPARENT_VERSION,
        _uuid_to_traceparent(trace_id),
        child_span_id,
        TRACE_FLAGS,
    )

    payload = {
        "trace_id": trace_id,
        "span_id": child_span_id,
        "parent_span_id": current_span_id,
        "service": service_name,
        "instance": instance_id,
        "clock": next_clock_vector,
    }

    raceway_clock = CLOCK_VERSION_PREFIX + _encode_base64url(json.dumps(payload, separators=(",", ":")))

    headers: Dict[str, str] = {
        TRACEPARENT_HEADER: traceparent,
        RACEWAY_CLOCK_HEADER: raceway_clock,
    }
    if tracestate:
        headers[TRACESTATE_HEADER] = tracestate

    return PropagationHeaders(headers=headers, clock_vector=next_clock_vector, child_span_id=child_span_id)


def increment_clock_vector(
    clock_vector: List[Tuple[str, int]],
    *,
    service_name: str,
    instance_id: str,
) -> List[Tuple[str, int]]:
    """Increment the clock component for this service/instance."""
    component = _clock_component(service_name, instance_id)
    updated: List[Tuple[str, int]] = []
    found = False
    for entry_component, value in clock_vector:
        if entry_component == component:
            updated.append((entry_component, value + 1))
            found = True
        else:
            updated.append((entry_component, value))

    if not found:
        updated.append((component, 1))

    return updated


def _parse_traceparent(value: str) -> Optional[Dict[str, str]]:
    parts = value.strip().split("-")
    if len(parts) != 4:
        return None

    _, trace_id_hex, span_id_hex, _ = parts
    if not _is_hex(trace_id_hex, 32) or not _is_hex(span_id_hex, 16):
        return None

    return {
        "trace_id": _traceparent_to_uuid(trace_id_hex),
        "span_id": span_id_hex.lower(),
    }


def _parse_raceway_clock(value: str) -> Optional[Dict[str, Optional[str]]]:
    if not value.startswith(CLOCK_VERSION_PREFIX):
        return None

    encoded = value[len(CLOCK_VERSION_PREFIX) :]
    try:
        payload = json.loads(_decode_base64url(encoded))
    except (json.JSONDecodeError, ValueError, binascii.Error):  # type: ignore[name-defined]
        return None

    clock_entries = []
    for item in payload.get("clock", []):
        if (
            isinstance(item, list)
            and len(item) == 2
            and isinstance(item[0], str)
            and isinstance(item[1], (int, float))
        ):
            clock_entries.append((item[0], int(item[1])))

    return {
        "trace_id": payload.get("trace_id"),
        "parent_span_id": payload.get("span_id"),  # span_id from sender becomes parent for receiver
        "clock": clock_entries,
    }


def _clock_component(service_name: str, instance_id: str) -> str:
    return f"{service_name}#{instance_id}"


def _generate_span_id() -> str:
    return secrets.token_hex(8)


def _uuid_to_traceparent(value: str) -> str:
    return value.replace("-", "").ljust(32, "0")[:32]


def _traceparent_to_uuid(value: str) -> str:
    return "{}-{}-{}-{}-{}".format(
        value[0:8],
        value[8:12],
        value[12:16],
        value[16:20],
        value[20:32],
    )


def _is_hex(value: str, expected_length: int) -> bool:
    if len(value) != expected_length:
        return False
    try:
        int(value, 16)
        return True
    except ValueError:
        return False


def _encode_base64url(value: str) -> str:
    return base64.urlsafe_b64encode(value.encode("utf-8")).decode("utf-8").rstrip("=")


def _decode_base64url(value: str) -> str:
    padding = "=" * ((4 - len(value) % 4) % 4)
    return base64.urlsafe_b64decode((value + padding).encode("utf-8")).decode("utf-8")
